### DISCUSSION TITLE
Custom OAuth 2.0 Provider with PKCE not working? (Laravel Passport)

### DISCUSSION THREAD

**Discussion by abbluiz** [nextauthjs/next-auth#2778] [Help] [answered]
### Question 💬

Hi, I am trying locally test OAuth 2.0 authentication through a local Laravel Passport server (OAuth 2.0) with PKCE. I am receiving the error "cookie code_verifier not found" when the authorization flow redirects back to nextauth app. Any ideas?

Authorization Code grant with PKCE is supported by Laravel Passport. Here's my provider configuration:

```javascript
const options = {
  providers: [
    {
      id: 'userh-account',
      name: 'UseRH Account',
      type: 'oauth',
      version: '2.0',
      scope: '',
      params: { grant_type: 'authorization_code' },
      accessTokenUrl: process.env.USERH_AUTHZ_SERVER_ORIGIN + '/oauth/token',
      requestTokenUrl: process.env.USERH_AUTHZ_SERVER_ORIGIN + '/oauth/token',
      authorizationUrl:
        process.env.USERH_AUTHZ_SERVER_ORIGIN +
        '/oauth/authorize?response_type=code',
      profileUrl: process.env.USERH_AUTHZ_SERVER_ORIGIN + '/api/user',
      async profile(profile, tokens) {
        return {
          ...profile,
          tokens
        }
      },
      clientId: process.env.USERH_AUTHZ_CLIENT_ID,
      protection: 'pkce',
    },
  ],

  session: {
    jwt: true,
  },

  jwt: {},

  pages: {},

  callbacks: {},

  events: {},

  theme: 'light',

  debug: true,
}

```

### How to reproduce ☕️

1. Authz server is running locally on port 80 and nextauth is running locally on port 3000
2. Nextauth redirects to authorization flow, I login through authz, redirects back to nextauth with "The code_verifier cookie was not found."

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by balazsorban44** [comment, marked as answer]
The error originates from here:

https://github.com/nextauthjs/next-auth/blob/e8a58a01b6b8c456268771f9b382fa300ccab8ac/src/server/lib/oauth/pkce-handler.js#L25

The `code_verifier` is saved in a secure cookie temporarily:

https://github.com/nextauthjs/next-auth/blob/e8a58a01b6b8c456268771f9b382fa300ccab8ac/src/server/lib/oauth/pkce-handler.js#L84-L87

Make sure that you have cookies enabled. Also, note that the `code_verifier` cookie has a lifetime of 15 minutes:

https://github.com/nextauthjs/next-auth/blob/e8a58a01b6b8c456268771f9b382fa300ccab8ac/src/server/lib/oauth/pkce-handler.js#L9

After that, the user has to try to log in again.

**Reply by abbluiz** [reply]
Thank you for your answer. I am still debugging this issue, however I know it's not the 15 minutes lifetime.

Is the authz server responsible to send this cookie back to nextauth?

**Reply by balazsorban44** [reply]
we set this cookie just before we redirect the user to the authorization endpoint, and we delete it after it has been used

**Reply by abbluiz** [reply]
So the authorization server must send the cookie back? Maybe the problem lies there

**Reply by balazsorban44** [reply]
See:

https://auth0.com/docs/authorization/flows/authorization-code-flow-with-proof-key-for-code-exchange-pkce

https://oauth.net/2/pkce/

**Reply by abbluiz** [reply]
Nevermind. This is an issue with Inertia.js on the Authorization Server (Larevel Passport + Laravel Fortify + Inertia). For anyone with the same issue, there are problems with Inertia redirecting to external URIs.

Thank you for your answers!
